File: src/uva/Leetcode00018.java
import java.util.*;

public class Leetcode00018 {
 
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        
        List<List<Integer>> ll = new ArrayList<>();
        Set<String> unique = new HashSet<>();
        
        for (int i = 0;i < nums.length;i++) {
            for (int j = i+1;j < nums.length;j++) {
                int k = j+1;
                int l = nums.length - 1;
                while (k < l) {
                    int sum = nums[i] + nums[j] + nums[k] + nums[l];
                    if (sum > target) {
                        l--;
                    } else if (sum < target) {
                        k++;
                    } else if (sum == target) {
                        List<Integer> v = Arrays.asList(nums[i], nums[j], nums[k], nums[l]);
                        if (unique.add(v.toString())) {
                            ll.add(v);
                        }
                        l--;
                        k++;
                    }
                }
            }
        }
        return ll;
    }
}

File: src/main/Leetcode00018.swift
class Solution {
    func fourSum(nums: [Int], _ target: Int) -> [[Int]] {
        var m = [String : [Int]]()
        let arr = nums.sort()
        for i in 0.stride(to: arr.count-3, by: 1) {
            for j in (i+1).stride(to: arr.count-2, by: 1) {
                var k = j + 1
                var l = arr.count - 1
                while k < l {
                    let sum = arr[i] + arr[j] + arr[k] + arr[l]
                    if sum == target {
                        let key = "(\(arr[i]), \(arr[j]), \(arr[k]), \(arr[l]))"
                        if m[key] == nil {
                            m[key] = [arr[i], arr[j], arr[k], arr[l]]
                        }
                        l -= 1
                        k += 1
                    } else if sum > target {
                        l -= 1
                    } else {
                        k += 1
                    }
                }
            }

        }
        return Array(m.values)
    }
}
